squire_lib/src/pairings/greedy.rs

```diff
@@ -1,58 +1,344 @@
+            (players[1], [players[0]].into_iter().collect()),
+        ]
+        .into_iter()
+        .collect();
+
+        let pairings = super::greedy_pairings(players.iter().cloned(), &opponents, 4, 0);
+        let Pairings {
+            mut paired,
+            rejected,
+        } = pairings;
+        assert_eq!(
+            paired.len(),
+            1,
+            "Exactly one pairing should have been produced"
+        );
+        assert_eq!(
+            rejected.len(),
+            1,
+            "Exactly one player should have been rejected from pairing"
+        );
+
+        assert_eq!(
+            &rejected[0], &players[1],
+            "The second player should have been rejected"
+        );
+
+        let pair = paired.remove(0);
+        drop(paired);
+        assert_eq!(pair.len(), 4, "Pairing should contain four players");
+        for player in pair {
+            assert!(
+                [players[0], players[2], players[3], players[4]].contains(&player),
+                "Pairing should only contain all except the second player"
+            )
+        }
+    }
+
+    #[test]
+    fn pair_pairs() {
+        let players: Vec<_> = std::iter::repeat_with(|| PlayerId::new(Uuid::new_v4()))
+            .take(8)
+            .collect();
+
+        let pairings = super::greedy_pairings(players.iter().cloned(), &HashMap::new(), 2, 0);
+        let Pairings { paired, rejected } = pairings;
+        assert_eq!(
+            paired.len(),
+            4,
+            "Exactly four pairing should have been produced"
+        );
+        assert_eq!(
+            rejected.len(),
+            0,
+            "No player should have been rejected from a pairing"
+        );
+
+        assert!(paired
+            .iter()
+            .any(|v| v.len() == 2 && v.contains(&players[0]) && v.contains(&players[1])));
+        assert!(paired
+            .iter()
+            .any(|v| v.len() == 2 && v.contains(&players[2]) && v.contains(&players[3])));
+        assert!(paired
+            .iter()
+            .any(|v| v.len() == 2 && v.contains(&players[4]) && v.contains(&players[5])));
+        assert!(paired
+            .iter()
+            .any(|v| v.len() == 2 && v.contains(&players[6]) && v.contains(&players[7])));
+
+        let opponents = paired
+            .into_iter()
+            .flat_map(|pair| [(pair[0], pair[1]), (pair[1], pair[0])])
+            .map(|(a, b)| (a, [b].into_iter().collect()))
+            .collect();
+        let pairings = super::greedy_pairings(players.iter().cloned(), &opponents, 2, 0);
+        let Pairings { paired, rejected } = pairings;
+        assert_eq!(
+            paired.len(),
+            4,
+            "Exactly four pairing should have been produced"
+        );
+        assert_eq!(
+            rejected.len(),
+            0,
+            "No player should have been rejected from a pairing"
+        );
+
+        assert!(paired
```

Comment:
Similar question here. Shouldn't the first pair be players zero and two (and so on)?